File: database/mybatis/mybatis-annotation/src/main/java/com/example/UserService.java
package com.example;

import static java.util.stream.Collectors.toList;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.List;

import org.apache.ibatis.cursor.Cursor;
import org.apache.ibatis.session.RowBounds;
import org.springframework.data.domain.Pageable;
import org.springframework.data.util.StreamUtils;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import lombok.RequiredArgsConstructor;

@Service
@Transactional
@RequiredArgsConstructor
public class UserService {
    private final UserMapper userMapper;

    public List<User> findAll(Pageable pageable) {
        return userMapper.findAll(toRowBounds(pageable));
    }

    public List<User> findAll() {
        Cursor<User> cursor = userMapper.findAllWithCursor();
        try (cursor) {
            return StreamUtils.createStreamFromIterator(cursor.iterator())
                              .collect(toList());
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    public User findById(long id) {
        return userMapper.findById(id).orElseThrow();
    }

    public void create(String name) {
        User account = new User();
        account.setName(name);
        userMapper.insert(account);
    }

    public void update(long id, String name) {
        User account = userMapper.findById(id).orElseThrow();
        account.setName(name);
        userMapper.update(account);
    }

    public void delete(long id) {
        userMapper.delete(id);
    }

    private static RowBounds toRowBounds(Pageable pageable) {
        return new RowBounds((int) pageable.getOffset(), pageable.getPageSize());
    }
}

File: kotlin/database/exposed/src/main/kotlin/com/example/UserService.kt
package com.example

import org.jetbrains.exposed.sql.batchInsert
import org.jetbrains.exposed.sql.deleteAll
import org.jetbrains.exposed.sql.deleteWhere
import org.jetbrains.exposed.sql.insert
import org.jetbrains.exposed.sql.select
import org.jetbrains.exposed.sql.selectAll
import org.jetbrains.exposed.sql.update
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class UserService {

    fun selectAll(): List<User> =
        Users.selectAll().map { User(it[Users.id], it[Users.name]) }

    fun select(id: Int): User =
        Users.select { Users.id eq id }.map { User(it[Users.id], it[Users.name]) }.single()

    fun insert(user: User) =
        Users.insert {
            it[name] = user.name
        }

    fun insert(users: List<User>) =
        Users.batchInsert(users) {
            this[Users.name] = it.name
        }

    fun update(user: User) =
        Users.update({ Users.id eq user.id }) {
            it[name] = user.name
        }

    fun deleteAll() = Users.deleteAll()

    fun delete(id: Int) = Users.deleteWhere { Users.id eq id }
}
